### ISSUE TITLE
Node support tracking issue

### ISSUE THREAD

**Issue by ryanto** [miragejs/miragejs#210]
This issue is to track the work needed for making Mirage's server runnable in Node. I want to write up our current thinking so that we can start to solicit feedback.

Mirage works by intercepting browser calls to `fetch` and `XMLHttpRequst` and returning a mocked HTTP response. For Node environments the approach we are taking is a bit different. Instead of intercepting all outgoing network requests from Node we will be providing a lower level APIs that allow you generate Mirage responses in user land.

For example, there will be a `server.handleRequest({ method: 'GET', path: '/articles', ... })` function to generate a Mirage response for a `GET /articles`.

The reason for this approach is that it allows you to use a Mirage response however you see fit. It can be served from an express server or used with a node mocking library like nock. Because the use cases for running Mirage in node are much more diverse than the browser, we feel providing these lower level APIs is a good first step.

The first APIs we will add to `server` are:

* `canHandle(request): boolean`
* `handleRequest(request): response`

At this time we are unsure about the shape of the request/response objects. We will most likely look to current Mirage/Pretender and use these objects as a starting point.

Any feedback would be greatly appreciated! Thanks

**Comment by samselikoff** [comment]
Great writeup!

As Ryan said, the shape of the request object is going to be something we'll need to figure out. Right now Mirage has a few places where we implicitly rely on Pretender's request object, but because running Mirage in node means making it more general than Pretender, we want to look in several places for inspiration on what object to standardize on (for example, Express's request object). We're still not sure how much of the request we need to expose as part of Mirage's APIs.

There's a lot of unknowns here, but in an effort to break this up into the smallest possible steps, we want to start with `canHandle(request)`.

The idea with `canHandle` is so that Mirage can be used in a situation like this:

```js
app.all('*', function (req, res, next) {
  if (mirageServer.canHandle({ method: req.method, url: req.url })) {
    let response = mirageServer.handle({ method: req.method, url: req.url, body: req.body })
    res.send(response)
  } else {
    next()
  }
})
```

Here, the user is adding an Express middleware that's letting Mirage handle a response, if it has a route defined. If not, the request can continue on down the stack. 

So, to kick off the node work, let's focus on `canHandle` first.

---

We have PR where we started this: https://github.com/miragejs/miragejs/pull/159

The work there is basically good, it really just needs to be brought across the finish line.

- There's a new `RequestHandler` class that `server` delegates to
- It uses RouteRecognizer (the underlying lib used by Pretender) to determine whether Mirage has a defined handler for a given route
- There are a few remaining tests we need to add listed in the Conversation tab of the PR
- To explain our test conventions, `external` is used for external APIs (meaning used by external users, actual users of Mirage) where `internal` is for everything else. Further, we have `external` split into browser-only, node-only, or shared (which are APIs that are supported in both browser and node environments).
- Given all that, `canHandle` is an external API that needs to work in both the browser and node, which is why its test is under `external/shared`

So, if someone wants to take this on, here's the next step I'd like to see. Open a new PR, that basically brings over the code from #159. But make sure you understand it, say what you think if anything new comes up, and add tests for the three other situations we laid out in the Conversation tab there.

---

Once we land canHandle, we can start talking about `handle` - which is the really exciting API!

**Comment by asantos00** [comment]
What about using Node's own request object?

I'm just suggesting this because for instance `express` extends that object, while `koa` uses it directly. By supporting it we can pretty much guarantee that is going to work anywhere, don't we?

Check out this `express` docs page: https://expressjs.com/en/api.html#req

> The req object is an enhanced version of Node’s own request object and supports all built-in fields and methods.

**Comment by samselikoff** [comment]
Reference code from an old PR: https://github.com/miragejs/miragejs/pull/29

**Comment by samselikoff** [comment]
Reference code on server.canHandle from old PR: https://github.com/miragejs/miragejs/pull/159

**Comment by martin-sweeny** [comment]
What's going on with this? Are you still open to contributions?

**Comment by samselikoff** [comment]
Definitely open to contributions! Unfortunately we had to shift focus away from OSS for now due to financial support reasons.

If you'd like to work on this, we left off over here: https://github.com/miragejs/miragejs/pull/699

I'd be happy to discuss the work that remains, if you're willing & able to invest the time on the feature!

**Comment by od-cyera** [comment]
Any update? also, do you know if the db + orm should work on node currently without changes? i know this is a bit off topic but depending on the answer i might be able to contribute a little :)

**Comment by samselikoff** [comment]
No update but almost positive the db/orm stuff doesn't depend on any browser apis. Biggest challenge is/was pretender.

**Comment by od-cyera** [comment]
Just checked, it does work, but when using typescript you cant import and use only the orm without the server. i noticed more issues with the types. might open a pr if i get the chance.. anyway, thanks for the quick reply!

**Comment by nomnomnomnom** [comment]
I wanted Node support, too, so I made a package which which mounts Mirage route handlers into an Express app: [mount-mirage-in-express](https://github.com/nomnomnomnom/mount-mirage-in-express).  (Redundant disclaimer: I made it.)

It's basically a small adapter which registers the Mirage handlers directly with Express and converts between the Express request/response and a Mirage request/response, so it works with the current Mirage as-is.
